drgn/internal/sudohelper.py

```diff
@@ -0,0 +1,88 @@
+# Copyright (c) Stephen Brennan <stephen@brennan.io>
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+"""Helper for opening a file as root and transmitting it via unix socket"""
+import array
+import os
+from pathlib import Path
+import pickle
+import socket
+import subprocess
+import sys
+import tempfile
+from typing import Union
+
+
+def recv_fd(sock: socket.socket) -> int:
+    """Receive a file descripter on the socket"""
+    # Borrows heavily from the Python docs for sock.recvmsg()
+    fds = array.array("i")
+    msg, ancdata, flags, addr = sock.recvmsg(4096, socket.CMSG_SPACE(fds.itemsize))
+    if msg == b"success" and ancdata:
+        level, typ, data = ancdata[0]
+        assert level == socket.SOL_SOCKET
+        assert typ == socket.SCM_RIGHTS
+        data = data[: fds.itemsize]
+        fds.frombytes(data)
+        return fds[0]
+    else:
+        try:
+            exc = pickle.loads(msg)
+        except Exception:
+            exc = Exception(msg.decode())
+        raise exc
+
+
+def send_fd(sock: socket.socket, fd: int) -> None:
+    """Send a file descriptor on the socket"""
+    # Borrows heavily from the Python docs for sock.sendmsg()
+    fds = array.array("i", [fd])
+    sock.sendmsg(
+        [b"success"],
+        [(socket.SOL_SOCKET, socket.SCM_RIGHTS, fds)],
+    )
```

Comment:
Similarly here, this is small enough to go directly in `main()` instead of a separate function.